### DISCUSSION TITLE
Question: Conversion in PFR

### DISCUSSION THREAD

**Discussion by kookma** [DanWBR/dwsim#193] [Q&A] [answered]
Is the conversion is PFR is calculated as below?

https://github.com/DanWBR/dwsim6/blob/66148417ad695b82065b53718709db313f8d40f3/DWSIM.UnitOperations/Reactors/PFR.vb#L647-L655

I am not sure if this is equal to:  `x= (Fa0 - Fa)/Fa0`

I appreciate if you kindly explain this like

```
ComponentConversions(sb.Name) = Abs(n0 - nf) / nf 
```

Furthermore, can I export conversion along the reactor and plot conversion profile?  I mean through Script Manager (Ironpython)

**Comment by DanWBR** [comment, marked as answer]
It should be

`ComponentConversions(sb.Name) = Abs(n0 - nf) / n0
`

n0 = initial mole flow in mol/s
nf = final mole flow in mol/s

**Reply by kookma** [reply]
Should I make an issue to correct this?

Side note:
If our group can contribute to reactor simulation, how we can proceed with this?
My proposal is:

- create a reactor branch
- let us to fork
- we will submit a proposal (a PPT will be prepared)
- get your approval
- start coding
This way we can proceed through a separate branch and we finally approve it you can merge it into the main branch!

**Important**: I propose a revamp and fundamental improvement to tubular/cstr reactors! Then we can move to conversion and equilibrium which are simpler case and lastly to gibbs! BUT our focal point is PFR.

**Reply by Greg66** [reply]
Hi,
yes, this procedure exactely is the intendet way to contribute to an open source projekt like DWSIM.

You should install Visual Studio (Community Edition) to debug your code prior to submission. [https://visualstudio.microsoft.com/de/vs/community/](url)

I also recommend to install GitHub-Desktop software. But it should also be possible to commit code to GitHub directely from Visual Studio.

Regards
Gregor

**Reply by DanWBR** [reply]
> Should I make an issue to correct this?

I've already corrected it in v6.6.2. That part is in the dynamic mode calculation routine, the steady state one was ok.

**Comment by DanWBR** [comment]
@kookma by the way, gibbs and equilibrium reactors are actually the hardest ones. PFR/CSTR and conversion are not that hard to model. 

You have many options if you want to "revamp" them. One is to do what you've described, another one would be to create a new, external unit operation, and the last one would be to create a custom model in IronPython or python.net, where you could use scipy's ODE solvers, for instance.

**Comment by kookma** [comment]
@Greg66 - thank you for your advice!

@DanWBR - python is the best option for us! but then converting to .NET and merging to core would be time consuming process!
Is there any document out there to see what is DWSIM conventions for coding? I assume we have to follow DWSIM coding rules, if we want to contribute to the core code!

**Comment by DanWBR** [comment]
@kookma there are no coding conventions at the moment. Why don't you create an external unit operation then? There is a full VB project as a reference here (https://github.com/DanWBR/DWSIM.UnitOperations.Humidifer) and a video tutorial from beginning to the end here: https://www.youtube.com/watch?v=anNCig-wgf4&list=PLWUTvgwTyRnOjZ5WtuWjwfc1G3GjeKTjI

You can also take a look at the Neural Network UO code in C#. https://github.com/DanWBR/DWSIM.UnitOperations.NeuralNetwork

**Reply by kookma** [reply]
Thank you Daniel! I was not aware of them. For sure I study them to see how we can proceed!
These tutorials are great!

The reason I said code conversion was to improve the current PFR written in VB.NET!

**Reply by DanWBR** [reply]
> Thank you Daniel! I was not aware of them. For sure I study them to see how we can proceed!
> These tutorials are great!
> 
> The reason I said code conversion was to improve the current PFR written in VB.NET!

You can then create a new Unit Operation (i.e. Advanced PFR) which inherits from the PFR class and then create a new Calculate() routine and upgrade the editor window, if required. Even easier and doesn't mess with the existing code.

**Reply by DanWBR** [reply]
The new UO can then be written either in VB or C#.

**Reply by kookma** [reply]
Thank you for your advices! We go as you recommended! My plan is to define few theses on this and I absolutely will consult you and get your opinion then!

**Reply by DanWBR** [reply]
Sure, good luck and let me know if you need help!

**Reply by DanWBR** [reply]
The advantage of inheriting an existing unit operation is that you will get everything from it, all its properties and the property editor itself. You can then define new properties if needed and override the calculation routine (Calculate()). It will work right out of the box.

**Comment by kookma** [comment]
@DanWBR 

can I export conversion along the reactor and plot conversion profile? I mean through Script Manager (Ironpython)
how about concentration/partial pressure of species?

**Reply by DanWBR** [reply]
the PFR has a **points** field that stores pressure, temperature and composition profile. The order of the component concentrations is the same as in ComponentConversions:

https://github.com/DanWBR/dwsim6/blob/aa13b15ef8894a15aab0d88c5fbff2e333b6aef1/DWSIM.UnitOperations/Reactors/PFR.vb#L618-L633

**Reply by kookma** [reply]
Many thanks Daniel!

**Reply by DanWBR** [reply]
@kookma points is an ArrayList where each element is a vector of doubles.

**Comment by kookma** [comment]
@DanWBR - I successfully extracted the data from a PFR in a python script and plot the profiles. 
I will submit the example as a Tiddlywiki file for DWSIM documentation for you! See the DWSIM simulation attached!

[plug-flow-reactor-plot-profiles-using-python.zip](https://github.com/DanWBR/dwsim6/files/7148041/plug-flow-reactor-plot-profiles-using-python.zip)  [This uses Numpy, Matplotlib with Python.NET]

**Some thoughts**

1. The points field stores, z (location), T, P and concentration of species. as described by Daniel above (https://github.com/DanWBR/dwsim6/discussions/193#discussioncomment-1306534)
2. If you want to calculate the conversion in Python you can use of the concentration derived from points field. The problem here is if you have a gas phase reaction with pressure drop you need flowrate and you cannot use concentration, because 0f the change of volume.
3. No data for volumetric flowrate and superficial velocity is stored.

**Suggestions**

1. Store flowrates instead of concentration (this addresses conversion calculation for gas phase reaction)
2. Store volumetric flowrate, this way one can calculates concentration if needed

from the above two data you can calculate:
- partial pressure, and mole fractions for gas phase reaction
- conversion  for both liquid and gas phase reaction
- superficial velocity (by dividing volumetric flowrate to reactor cross sectional area)
- concentration by dividing molar flowrates by volumetric flow

So points shall have `z`, `F` molar flowrates of all species, `T`, `P`, `v`
`v` stands for volumetric flowrate.

If you like you keep also concentration as before but molar flowrate is enough!
